x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -306,7 +314,108 @@ private ModelConfigMap createModelConfigMap(SearchHits hits, String inferenceEnt
+
+        String inferenceEntityId = newModel.getConfigurations().getInferenceEntityId();
+        logger.info("Attempting to store update to inference endpoint [{}]", inferenceEntityId);
+
+        if (preventDeletionLock.contains(inferenceEntityId)) {
+            logger.warn(format("Attempted to update endpoint [{}] that is already being updated", inferenceEntityId));
+            finalListener.onFailure(
+                new ElasticsearchStatusException(
+                    "Endpoint [{}] is currently being updated. Try again once the update completes",
+                    RestStatus.CONFLICT,
+                    inferenceEntityId
+                )
+            );
+            return;
+        } else {
+            preventDeletionLock.add(inferenceEntityId);
+        }
+
+        SubscribableListener.<BulkResponse>newForked((subListener) -> {
+
+            IndexRequest configRequest = createIndexRequest(
+                Model.documentId(inferenceEntityId),
+                InferenceIndex.INDEX_NAME,
+                newModel.getConfigurations(),
+                true
+            );
+
+            client.prepareBulk().add(configRequest).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).execute(subListener);
+
+        }).<BulkResponse>andThen((subListener, configResponse) -> {
+
+            if (configResponse.hasFailures()) {
+                logger.error(
+                    format("Failed to update inference endpoint [%s] due to [%s]", inferenceEntityId, configResponse.buildFailureMessage())
+                );
+                finalListener.onFailure(
+                    new ElasticsearchStatusException(
+                        format("Failed to update inference endpoint [%s] due to [%s]", inferenceEntityId),
+                        RestStatus.INTERNAL_SERVER_ERROR,
+                        configResponse.buildFailureMessage()
+                    )
+                );
+            } else {
+                IndexRequest secretsRequest = createIndexRequest(
+                    Model.documentId(newModel.getConfigurations().getInferenceEntityId()),
+                    InferenceSecretsIndex.INDEX_NAME,
+                    newModel.getSecrets(),
+                    true
+                );
+
+                client.prepareBulk().add(secretsRequest).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).execute(subListener);
+            }
+        }).<BulkResponse>andThen((subListener, secretsResponse) -> {
+            if (secretsResponse.hasFailures()) {
+                IndexRequest configRequest = createIndexRequest(
+                    Model.documentId(inferenceEntityId),
+                    InferenceIndex.INDEX_NAME,
+                    existingModel.getConfigurations(),
+                    true
+                );
+
+                client.prepareBulk().add(configRequest).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).execute(subListener);
+                logger.error(
```

Comment:
Does this call the below attached listener if we call `subListener.onFailure`? If yes, I think we'll log two errors for the same issue.  If no, I think we need to `preventDeletionLock.remove(inferenceEntityId);`

Might be easier to always call `subListener` in each block, then replace that last block with:
```
.addListener(finalListener.delegateResponse(subListener, e) -> {
            preventDeletionLock.remove(inferenceEntityId);
            finalListener.onFailure(
                new ElasticsearchStatusException(
                    format(
                        "Failed to rollback while handling failure to update inference endpoint [%s]. "
                            + "Endpoint may be in an inconsistent state due to [%s]",
                        inferenceEntityId
                    ),
                    RestStatus.INTERNAL_SERVER_ERROR,
                    e
                )
            );
        });
```

Reply:
Subscribable listener documentation says "A failure of any step will bypass the remaining steps and ultimately fail finalListener.". I do wonder if these lines after 379 will ever actually be executed... Actually, I think they need to be added into the listener passed on 379.

I think that the reason that the final block is set up like this is because the bulk API doesn't throw exceptions, it catches them into this `BulkResponse configResponse` object, and we need to turn that into an exception for the listener.

Similarly, I am calling onResponse or onFailure for the finalListener instead of the subListener because I need to call those functions with the correct object types (boolean). The last block isn't really for the whole set of calls, its only for catching the case

Reply:
Actually, it also occurs to me that if the bulk API throws an exception for some reason, we won't properly remove the lock, so I need to add a new action listener